homeassistant/components/tado/device_tracker.py

```diff
@@ -1,149 +1,141 @@
-        """Return the name of the given device or None if we don't know."""
-        filter_named = [
-            result.name for result in self.last_results if result.mac == device
-        ]
-
-        if filter_named:
-            return filter_named[0]
-        return None
-
-    @Throttle(MIN_TIME_BETWEEN_SCANS)
-    async def _async_update_info(self):
-        """Query Tado for device marked as at home.
-
-        Returns boolean if scanning successful.
-        """
-        _LOGGER.debug("Requesting Tado")
-
-        if self.websession is None:
-            self.websession = async_create_clientsession(
-                self.hass, cookie_jar=aiohttp.CookieJar(unsafe=True)
+            _LOGGER.debug("Tado device %s is not at home", device["name"])
+
+        self.async_write_ha_state()
+
+    async def async_added_to_hass(self) -> None:
+        """Register state update callback."""
+        _LOGGER.debug("Registering Tado device tracker entity")
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass,
+                SIGNAL_TADO_MOBILE_DEVICE_UPDATE_RECEIVED,
+                self.update_state,
             )
+        )
 
-        last_results = []
-
-        try:
-            async with asyncio.timeout(10):
-                # Format the URL here, so we can log the template URL if
-                # anything goes wrong without exposing username and password.
-                url = self.tadoapiurl.format(
-                    home_id=self.home_id, username=self.username, password=self.password
-                )
-
-                response = await self.websession.get(url)
-
-                if response.status != HTTPStatus.OK:
-                    _LOGGER.warning("Error %d on %s", response.status, self.tadoapiurl)
-                    return False
-
-                tado_json = await response.json()
-
-        except (asyncio.TimeoutError, aiohttp.ClientError):
-            _LOGGER.error("Cannot load Tado data")
-            return False
-
-        # Without a home_id, we fetched an URL where the mobile devices can be
-        # found under the mobileDevices key.
-        if "mobileDevices" in tado_json:
-            tado_json = tado_json["mobileDevices"]
-
-        # Find devices that have geofencing enabled, and are currently at home.
-        for mobile_device in tado_json:
-            if mobile_device.get("location") and mobile_device["location"]["atHome"]:
-                device_id = mobile_device["id"]
-                device_name = mobile_device["name"]
-                last_results.append(Device(device_id, device_name))
-
-        self.last_results = last_results
+    @property
+    def name(self) -> str:
+        """Return the name of the device."""
+        return self._device_name
 
-        _LOGGER.debug(
-            "Tado presence query successful, %d device(s) at home",
-            len(self.last_results),
-        )
+    @property
+    def is_connected(self) -> bool:
```

Comment:
Please also implement the mac address property and potentially hostname and/or ip address if possible.

https://developers.home-assistant.io/docs/core/entity/device-tracker#scannerentity

Reply:
This is the data returned from the API:

```
{
  name: "MySecretName",
  email: "mail@mail.com",
  username: "mail@mail.com",
  id: 1a76161-eb92-49c9-9550-853859ab8a03",
  homes: [
    {
      id: 1208159,
      name: "SECRET"
    }
  ],
  locale: "nl",
  mobileDevices: [
    {
      name: "Secret",
      id: 123456789,
      settings: {
      geoTrackingEnabled: false,
      specialOffersEnabled: false,
      onDemandLogRetrievalEnabled: false,
      pushNotifications: {
      lowBatteryReminder: true,
      awayModeReminder: true,
      homeModeReminder: true,
      openWindowReminder: true,
      energySavingsReportReminder: true,
      incidentDetection: true,
      energyIqReminder: false
    }
  },
  deviceMetadata: {
    platform: "Android",
    osVersion: "13",
    model: "Samsung_SM-G990B2",
    locale: "nl"
  }
}
]
}
```
A MAC address, according to the definition, is not defined _per mobile device_. Note the Mo We could use the generic ID, which is an integer. Same goes for hostname, I could use the general name of the mobile device. Yet, I didn't classify them as matching the definition. Would it be mandatory to still fill it in?

Reply:
Hmm. What kind of state data does the API return? We're accessing the `location` field. What can that return besides `atHome`? I'm thinking if we should change base class to TrackerEntity since we normally use that for GPS trackers.

Reply:
If the GPS is turned on, only then the location object is made visible. Hence, the check on `get("location")`. See the output when I turn on my GPS on my smartphone.

![image](https://github.com/home-assistant/core/assets/5011203/99f2687b-bc89-48b1-b635-8886333a392d)

I don't think a TrackerEntity would be suitable. Since it misses the longitude and latitude in the output, which are required for the TrackerEntity.
The `atHome` has also been used in the outdated DeviceTracker to mark it as active or inactive. Any new insights we can derive from this information?

Reply:
The TrackerEntity also supports returning the location name directly. I'm thinking if we should just use that and return `STATE_HOME` or `STATE_NOT_HOME`. The coordinates are not actually required for TrackerEntity. The developer docs aren't really accurate there.

I think this seems to be the closest to what the API reports. If the location data isn't present we should return `None` to represent unknown state.

Reply:
Use the device id to set a unique id for the entity.

Reply:
Both patched. Feel free to review.